src/mem/cache/tags/base_set_assoc.hh

```diff
@@ -163,19 +169,25 @@ class BaseSetAssoc : public BaseTags
      * @param is_secure True if the target memory space is secure.
      * @param size Size, in bits, of new block to allocate.
      * @param evict_blks Cache blocks to be evicted.
+     * @param partition_ID Partition ID for resource management.
      * @return Cache block to be replaced.
      */
     CacheBlk* findVictim(Addr addr, const bool is_secure,
                          const std::size_t size,
-                         std::vector<CacheBlk*>& evict_blks) override
+                         std::vector<CacheBlk*>& evict_blks,
+                         const uint64_t partition_ID=0) override
     {
         // Get possible entries to be victimized
-        const std::vector<ReplaceableEntry*> entries =
+        std::vector<ReplaceableEntry*> entries =
             indexingPolicy->getPossibleEntries(addr);
 
+        // Filter entries based on PartitionID
+        for (auto partitioning_policy : partitioningPolicies)
+            partitioning_policy->filterByPartition(entries, partition_ID);
+
         // Choose replacement victim from replacement candidates
-        CacheBlk* victim = static_cast<CacheBlk*>(replacementPolicy->getVictim(
-                                entries));
+        CacheBlk* victim = entries.empty() ? nullptr :
```

Comment:
What happens if the victim is nullptr? I don't think we handle this case. The findVictim() call should always return a victim from what I understand.

Reply:
From what I saw in `BaseCache::updateCompressionData()` and `BaseCache::allocateBlock()`, when the `findVictim()` function is called, the `victim` returned is always checked for whether it's a valid entry as so: 

```
//function BaseCache::updateCompressionData()
    if (!victim) {
        return false;
    }
...
//function BaseCache::allocateBlock()
    if (!victim)
        return nullptr;
```
This check should pick up a `nullptr` being returned by `findVictim()` as proposed in this PR

Reply:
Looking at the code for `replacementPolicy->getVictim(entries)` it asserts that the list of available entries is not empty - this in principle is not guaranteed when using any partitioning policy, as an upstream requestor may be allocated no capacity for the cache in question. 

This means that either every policy needs to ensure that every upstream requestor is allocated at least one way from each set of a cache (which would not be reflective of the intended functionality) or to continue relying on the existing check.

Reply:
The issue is not that RPs require candidates to have victims. The point that I am concerned about is returning nullptr as the selected victim. This path has not been tested in years, and I have no clue what happens in that scenario. My concern is entering some kind of deadlock state because no victim is found. I think that as long as you made sure to test it with a couple of simulation combinations (to make sure it does not get stuck), and allow the PPs vector to be empty (in order to keep the current behavior for those who do not want a PP), it should probably be fine to merge it as is.

Reply:
The way PPs are currently implemented allows for the partitioning logic to be completely ignored if the user does not provide any Partitioning configurations. The current code has been mainly tested with caches using `BaseSetAssociative` tags, which is why I've gone ahead and added a warning for the other Tag types. The provided config (`./configs/cache_partitioning/cache_partitioning.py`) was used to test the most probable allocations that an user is likely to make.